.github/cloudbuild/android.yaml

```diff
@@ -0,0 +1,111 @@
+steps:
+# We need this step to correctly identify dockerfile tag
+- id: fetch-repo-history
+  name: gcr.io/cloud-builders/git
+  timeout: 10s
+  args:
+  - fetch
+  - --unshallow
+# Download DOCKERFILE_COMMIT tag if it exists to skip docker image generation,
+#   or download the :latest tag and use it as a cache,
+#   or skip downloading if :latest doesn't exist yet.
+# This being a separate step helps readability in Cloud Build console.
+- id: cache-old-image
+  name: gcr.io/cloud-builders/docker
+  timeout: 600s # 10 minutes
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)

```

Comment:
>if it has a COPY instruction in the feature, then the image will be changed if the copied files are changed

Using a different tag alone but still using a cache from previous commit wouldn't solve this issue.
We would solve it only if we wouldn't use any cache at all, but in this case we would spend ~30 minutes for each build just building an image.

Also, while tagging current image with current hash would work (and I was doing so in the beginning), since the image itself doesn't change, docker sees that we push the same image, and just adds a new tag to it. So we would end up with an image which has millions of tags, one for each commit that ever went to our CI.
I'm not sure if docker has a limitation on how many tags an image can have, but this certainly hurts usability/readability.

Reply:
And by the way, I don't see too much uses for the `COPY` command.
Maybe we could use it for injecting the dependencies that can't be downloaded via the internet, but in this case we couldn't build the image on CI at all, so we would need to remove the step that's building the image altogether, and build the image manually when we need it. Not too convenient, so I don't think we would do such a thing on CI, so that's not relevant to this cloudbuild file, in my opinion.

Reply:
>Not sure if I understand your point, but the issue I mean is that we will have the same tag for 2 different images because docker build will detect the changed files and build a new image but tag it with the old commit. But I think it will only a problem if we need to debug the container.

The way it is written right now:
1. If there are no changes, we will pull the old image with the old tag, use the old image as a cache and do nothing.
2. If there were some changes, we will try to pull the image with the new hash, fail because there is no such image, and pull the `:latest` image in hope that we can use the cache at least partially.

This breaks if we use `COPY` to copy changed local files without changed in the Dockerfile.
In all other cases we have changed Dockerfile with changed hash.

...

Oh, I think now I understand what you are talking about.
No, it's impossible to have a modified Dockerfile with the old cache on CI, because each time the CI starts it gets a fresh repo without any changed.
The only cache we have is the one we explicitly download.
So the cleanup on CI is meaningless.

>In any case, we should have a cleanup step to delete old images / tags / artifacts.

How would you implement this?
If you are talking about the local state of the repository, then there is no need for this because Google does this for us.

>I used to have some patch defined in a file committed into the repo, then copy it into the image in docker build step. The first step is git pull so we have all the files in the repo available.

You don't have to do this when working with the Cloud Build. You already have your repo mounted inside your container.

And I would argue that a Dockerfile that does something like this is an invalid Dockerfile, and such things must be copied in the runtime, and not when you build the image.
So if your file is in another repo, that is not mounted by default, you can download it in a build step, without changing the docker image.

Reply:
Not following all of the above but as a goal we want:

1. Fast builds: Leverage build cache between CI builds to reduce build times, parallel builds (use of Make currently serializes tasks), use of ccache or other caching framework, do not re-generate docker image each time (validating image is OK) 
2. Clean builds: Remove output other than the bazel cache (maybe bazel cache should not live under output?) before new CI job
3. Debugability: If there was an error can I see the created artifacts?
4. Code quality: Enforce formatting, file format (e.g. line endings), etc, we should use clang-tidy, and flutter analyze to check the code (maybe not in CI but as a regular check of quality)

Reply:
This thread is already lengthy now. So would just want to summarize what I think:
1. Using commit hash of the `Dockerfile` instead of the repo may cause issue if the Dockerfile have `ADD` or `COPY` instruction in it. But this specific `Dockerfile` does not have them, so it will be fine now. It's your choice to keep using it or not.
2. In the cleanup step, **persistent** data (docker image stored in cloud registry, build artifacts stored in the cloud bucket) should be deleted if we want to avoid keeping them forever (exception is release build). Strategy can be time-based (keep the last n months) or count-based (keep the last n artifacts, images) or both.